pkg/cloud/openstack/machine/actuator.go

```diff
@@ -94,51 +99,45 @@ func (oc *OpenstackClient) Create(ctx context.Context, cluster *clusterv1.Cluste
 		return nil
 	}
 
-	// get machine startup script
-	var ok bool
-	userData := []byte{}
-	if providerSpec.UserDataSecret != nil {
-		namespace := providerSpec.UserDataSecret.Namespace
-		if namespace == "" {
-			namespace = machine.Namespace
-		}
-
-		if providerSpec.UserDataSecret.Name == "" {
-			return fmt.Errorf("UserDataSecret name must be provided")
-		}
+	userData, err := getUserDataSecret(kubeClient, providerSpec, machine.Namespace)
+	if err != nil {
+		return oc.handleMachineError(machine, apierrors.CreateMachine(
+			"error creating Openstack instance: %v", err))
+	}
 
-		userDataSecret, err := kubeClient.CoreV1().Secrets(namespace).Get(providerSpec.UserDataSecret.Name, metav1.GetOptions{})
+	isMaster := util.IsControlPlaneMachine(machine)
+	var userDataRendered string
+	if len(userData) > 0 {
+		klog.Info("")
+		var err error
+		userDataRendered, err = oc.getMachineScript(isMaster, cluster, machine, string(userData))
 		if err != nil {
-			return err
+			return oc.handleMachineError(machine, apierrors.CreateMachine(
+				"error creating Openstack instance: %v", err))
 		}
-
-		userData, ok = userDataSecret.Data[UserDataKey]
-		if !ok {
-			return fmt.Errorf("Machine's userdata secret %v in namespace %v did not contain key %v", providerSpec.UserDataSecret.Name, namespace, UserDataKey)
+	} else {
+		var distri string
+		if strings.Contains(strings.ToLower(providerSpec.Image), "ubuntu") {
+			distri = "ubuntu"
+		} else if strings.Contains(strings.ToLower(providerSpec.Image), "centos") {
+			distri = "centos"
```

Comment:
do we request the Image's name need to contain 'centos' keyword in order to work? this seems expect too much on user's input

Reply:
Currently yes.

I was hoping, that this will create as little user's input as possible. Adding another field to be filled for a machine... like e.g. `distributionType: ...`. But IMHO this causes more possibility to do an user error. You could specify an Ubuntu image, but use `distributionType: centos`. I thought tying the Distri-Information to the image would prevent this. What does others think on this @ainmosni , @flaper87 ?

For sure, I need to change the documentation for this.

Reply:
But. It just came to my mind... as soon as we support volumes, we need to specify the distributionType... So I will add it.